### PR TITLE
feat: configurable log formatter

### PR DESCRIPTION
**Pull request by leseb**
Using the new `log_format` option under the `general` section of the
config file, we can now configure an alternative log format that suits
our needs.The default line remains the same, however if you want to
remove timestamps from the log message you should remove `%(asctime)s`.

Note: the formating is validated, but make sure you read the module
[attributes](https://docs.python.org/3/library/logging.html#logrecord-attributes)
when building the format string.

Closes: https://github.com/instructlab/instructlab/issues/1720
Signed-off-by: Sébastien Han <seb@redhat.com>

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: .spellcheck-en-custom.txt [hunk 1]
```diff
@@ -1,164 +1,165 @@
+AMDGPU
 Abhishek
 Akash
-AMDGPU
-arge
+Bhandwaldar
```
Comment (leseb) [inline .spellcheck-en-custom.txt:4]: Looks like `make spellcheck-sort` didn't run for quite some time.

File: src/instructlab/configuration.py [hunk 2]
```diff
@@ -627,10 +631,14 @@ def init(
...
             config_obj.general.debug_level = debug_level
 
+        if log_format != log.FORMAT:
+            config_obj.general.log_format = log_format
+
         # setup logging
         log.configure_logging(
             log_level=config_obj.general.log_level.upper(),
             debug_level=config_obj.general.debug_level,
+            fmt=log_format,
```
Comment (bjhargrave) [inline src/instructlab/configuration.py:641]:
```suggestion
            fmt=config_obj.general.log_format,
```

Like the prior line, shouldn't this come from the config object?

File: src/instructlab/configuration.py [hunk 3]
```diff
@@ -627,10 +631,14 @@ def init(
             config_obj.general.log_level = "DEBUG"
             config_obj.general.debug_level = debug_level
 
+        if log_format != log.FORMAT:
```
Comment (bjhargrave) [inline src/instructlab/configuration.py:634]: Do we need this if? Why not always set the config log_format member?

File: src/instructlab/configuration.py (hunk 2 above)
Comment (leseb) [reply src/instructlab/configuration.py:648]: I changed the code, PTAL again, first we check for the flag, then try the config, then the default. Thanks

**Review by tiran — CHANGES_REQUESTED** [review]
IMO `--log-format` should be a parameter of `ilab config init`, not `ilab` root. This would allow users to initialize a config with custom logger configuration. I doubt that users are going to pass a long format string to every `ilab` call.

**Comment by leseb** [comment]
> IMO `--log-format` should be a parameter of `ilab config init`, not `ilab` root. This would allow users to initialize a config with custom logger configuration. I doubt that users are going to pass a long format string to every `ilab` call.

What if users don't have a config? AFAIK `config.yaml` is optional most of the time, for example, `ilab model serve` does not need it.

**Comment by tiran** [comment]
In the absence of a `config.yaml`, `ilab model serve` behaves like `ilab --config=DEFAULT model serve`. I think we don't have a design and UX doc for CLI and config, yet. Are we going to add CLI overrides for all config values? How would I generate a `config.yaml` with a different default log format?

**Comment by leseb** [comment]
> In the absence of a `config.yaml`, `ilab model serve` behaves like `ilab --config=DEFAULT model serve`. I think we don't have a design and UX doc for CLI and config, yet. Are we going to add CLI overrides for all config values? 

One of our goals is to integrate well with containers and Kubernetes. Using flags is much easier than generating and consuming a config file.

> How would I generate a `config.yaml` with a different default log format?

What do you mean? You can use:

```
general:
  log_format: <string>
```

Am I missing something?

**Comment by tiran** [comment]
> What do you mean? You can use:
> 
> ```
> general:
>   log_format: <string>
> ```
> 
> Am I missing something?

I would like to set the `log_format` value in the `ilab config init` call, so I can automatically generate a `config.yaml` with my desired log format.

**Comment by leseb** [comment]
> > What do you mean? You can use:
> > ```
> > general:
> >   log_format: <string>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Am I missing something?
> 
> I would like to set the `log_format` value in the `ilab config init` call, so I can automatically generate a `config.yaml` with my desired log format.

I just added support for that.

File: src/instructlab/log.py [hunk 4]
```diff
@@ -37,26 +39,28 @@ def isatty(self):
...
         file_handler = logging.FileHandler(log_file, mode, encoding="utf-8")
         logger.addHandler(file_handler)
         # Create formatter and set it for both handlers
-        formatter = CustomFormatter(FORMAT)
+        formatter = CustomFormatter(fmt)
         file_handler.setFormatter(formatter)
 
-    sys.stdout = LoggerWriter(logger.info)
-    sys.stderr = LoggerWriter(logger.error)
+    sys.stdout = typing.cast(typing.TextIO, LoggerWriter(logger.info))
```
Comment (booxter) [inline src/instructlab/log.py:55]: I don't understand why this type-cast is needed here. Could you please enlighten me?

File: src/instructlab/log.py [hunk 5]
```diff
@@ -67,7 +71,15 @@ def configure_logging(
...
-    stream.setFormatter(CustomFormatter(fmt=fmt))
+    try:
+        stream.setFormatter(CustomFormatter(fmt=fmt))
+    except ValueError as e:
+        print(
+            f"Failed to configure log format: {e}\n"
+            "Did you specified a valid log format?\n"
+            "Consider reading: https://docs.python.org/3/library/logging.html#logrecord-attributes"
+        )
+        sys.exit(1)
```
Comment (booxter) [inline src/instructlab/log.py:82]: Please don't explicitly exit from inside a library function. (It makes it hard to e.g. test it, or reuse in a context where exit is not what we want.) Instead, if you'd like to make `init` fail, propagate the error from `configure_logging` (either by returning a boolean flag or raising an exception), then handle it by the caller.

(That said, I am not sure we should have the `init` option for `log-format` at all, so perhaps the point is moot.)

File: src/instructlab/configuration.py [hunk 6]
```diff
@@ -632,10 +636,23 @@ def init(
             config_obj.general.log_level = "DEBUG"
             config_obj.general.debug_level = debug_level
 
+        # --log-format option overrides config file
+        # If the flag is not used
+        # Only do this, if not called from "config" or "init" (support the legacy `ilab init` command)
+        if ctx.invoked_subcommand not in ("config", "init"):
```
Comment (booxter) [inline src/instructlab/configuration.py:642]: how useful is it to pass a custom logger via cli? can't we do the same by modifying a config file - or passing a custom one, if really needed?

File: tests/test_config.py [hunk 7]
```diff
@@ -241,8 +241,19 @@ def test_config_modified_settings(self, tmp_path_home):
...
+    assert logging.getLogger("root").handlers[0].formatter._fmt == "%(message)s"
+
+
+def test_invalid_log_formatter(capfd):
+    with pytest.raises(SystemExit):
+        configure_logging(log_level="INFO", debug_level=0, fmt="invalid_log_format")
+    out, err = capfd.readouterr()
+    assert (
+        "Failed to configure log format" in out
+        or "Failed to configure log format" in err
```
Comment (booxter) [inline tests/test_config.py:258]: why `or` here? don't we know that we log to stdout and not stderr?

**Comment by booxter** [comment]
> Using flags is much easier than generating and consuming a config file.

In k8s world, this can be achieved by a `ConfigMap` that is [mounted](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#add-configmap-data-to-a-volume) into a container as a file. This seems to be a better UI than modifying CLI arguments when a different log format is needed.

I am generally suspicious of adding an option per config file attribute. I understand there's no UX standard defined for config files, yet, but if I were to choose one, I would make `ilab init` barebones, without per-attribute fields; then let the user modify the yaml config as they see fit; finally, validate the read config as per schema (defined in `pydantic`).

**Review by booxter — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -38,6 +38,11 @@
 * The `instructlab` package now uses optional dependencies for each supported hardware `cpu`,
   `cuda`, `hpu`, `mps`, and `rocm`. To install InstructLab for e.g. NVIDIA CUDA, use
   `pip install instructlab[cuda]`.
+* Introduces CUDA-specific requirements in a new file `requirements-cuda.txt`. To install these
```
[the remark this answers is not in the document]
Comment (leseb) [reply CHANGELOG.md:41]: Looks like it.

File: src/instructlab/log.py (hunk 4 above)
Comment (leseb) [reply src/instructlab/log.py:55]: mypy used to complain:

```
  Error: src/instructlab/log.py:53: error: Incompatible types in assignment (expression has type "LoggerWriter", variable has type "TextIO")  [assignment]
  Error: src/instructlab/log.py:54: error: Incompatible types in assignment (expression has type "LoggerWriter", variable has type "TextIO")  [assignment]
```

It's strange given that `LoggerWritter` implements the necessary methods to mimic a file-like object. Let me try again.

**Comment by leseb** [comment]
@tiran @booxter I drastically reduced the scope since some parts need more consensus. Now we only have a new `log_format` field in the configuration file. PTAL. Thanks!

File: src/instructlab/configuration.py (hunk 6 above)
Comment (leseb) [reply src/instructlab/configuration.py]: resolved with https://github.com/instructlab/instructlab/pull/1741#issuecomment-2252354702

File: src/instructlab/log.py (hunk 5 above)
Comment (leseb) [reply src/instructlab/log.py]: Resolved.

File: src/instructlab/log.py (hunk 4 above)
Comment (leseb) [reply src/instructlab/log.py]: Resolved with https://github.com/instructlab/instructlab/pull/1741#issuecomment-2252354702

File: src/instructlab/configuration.py [hunk 9]
```diff
@@ -606,7 +610,7 @@ def init(
     elif os.path.isfile(config_file):
         try:
             config_obj = read_config(config_file)
-        except ConfigException as e:
+        except (ConfigException, yaml.constructor.ConstructorError) as e:
```
[the remark this answers is not in the document]
Comment (leseb) [reply src/instructlab/configuration.py]: Removed for another PR.

File: src/instructlab/config/init.py [hunk 10]
```diff
@@ -60,6 +60,12 @@
     "using the same taxonomy repository. ",
 )
 @click.option("--train-profile", type=click.Path(), default=None)
+@click.option(
```
[the remark this answers is not in the document]
Comment (leseb) [reply src/instructlab/config/init.py]: Resolved https://github.com/instructlab/instructlab/pull/1741#issuecomment-2252354702

File: tests/test_config.py (hunk 7 above)
Comment (leseb) [reply tests/test_config.py]: Addressed.

File: src/instructlab/log.py [hunk 11]
```diff
@@ -67,7 +70,14 @@ def configure_logging(
...
     stream = logging.StreamHandler()
-    stream.setFormatter(CustomFormatter(fmt=fmt))
+    try:
+        stream.setFormatter(CustomFormatter(fmt=fmt))
+    except ValueError as e:
+        raise ValueError(
+            f"Failed to configure log format: {e}\n"
+            "Have you specified a valid log format?\n"
+            "Consider reading: https://docs.python.org/3/library/logging.html#logrecord-attributes"
+        ) from e
```
Comment (tiran) [inline src/instructlab/log.py:80]: This validation is way too deep in the stack. You should validate the parameter in click or in the config loader instead.

```python
>>> import logging
>>> logging.PercentStyle("invalid").validate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.11/logging/__init__.py", line 438, in validate
    raise ValueError("Invalid format '%s' for '%s' style" % (self._fmt, self.default_format[0]))
ValueError: Invalid format 'invalid' for '%' style
>>> logging.PercentStyle("%(message)s").validate()
```

File: tests/test_config.py [hunk 12]
```diff
@@ -313,3 +314,10 @@ def test_compare_default_config_testdata(
         "intentional, run 'make regenerate-testdata' and commit the "
         "updated test data."
     )
+
+
+def test_invalid_log_formatter(capfd):
+    with pytest.raises(ValueError):
+        configure_logging(log_level="INFO", debug_level=0, fmt="invalid_log_format")
+    out, _ = capfd.readouterr()
+    assert "Failed to configure log format" in out
```
Comment (tiran) [inline tests/test_config.py:323]:
```suggestion
    with pytest.raises(ValueError) as e:
        configure_logging(log_level="INFO", debug_level=0, fmt="invalid_log_format")
   e.match("Failed to configure log format")
```

**Review by tiran — CHANGES_REQUESTED** [review]

File: src/instructlab/config/init.py [hunk 13]
```diff
@@ -69,6 +76,7 @@ def init(
     repository,
     min_taxonomy,
     train_profile,
+    log_format: str = log.FORMAT,
```
[the remark this answers is not in the document]
Comment (leseb) [reply src/instructlab/config/init.py]: Irrevelant now, since this code was removed.

File: src/instructlab/configuration.py [hunk 14]
```diff
@@ -587,7 +588,10 @@ def ensure_config(self, ctx: click.Context) -> None:
 
 
 def init(
-    ctx: click.Context, config_file: str | os.PathLike[str], debug_level: int = 0
+    ctx: click.Context,
+    config_file: str | os.PathLike[str],
+    debug_level: int = 0,
+    log_format: str = "",
```
[the remark this answers is not in the document]
Comment (leseb) [reply src/instructlab/configuration.py]: Irrevelant now, since this code was removed.

File: tests/test_config.py [hunk 15]
```diff
@@ -177,6 +177,15 @@ def test_validate_log_level_valid(self):
...
         assert cfg.general.validate_log_level("NOTSET") == "NOTSET"
 
+    def test_validate_log_format_invalid(self):
+        cfg = config.get_default_config()
+        with pytest.raises(ValueError):
+            cfg.general.validate_log_format("INVALID")
+
+    def test_validate_log_format_valid(self):
+        cfg = config.get_default_config()
+        assert cfg.general.validate_log_format("%(message)s") == "%(message)s"
```
Comment (booxter) [inline tests/test_config.py:187]: how does it work? you don't return anything.

**Review by booxter — CHANGES_REQUESTED** [review]

**Comment by mairin** [comment]
This is proposed as a blocker - wondering what the impact is if it does not make 0.18.0? It will be more difficult to get good logs and help users who have difficulty running it...?

**Comment by booxter** [comment]
@mairin I don't think this should be 0.18 blocker. This allows to adjust log format (include or remove some details.) The default format is good enough I think and definitely has enough info to debug issues. I think the primary concern is probably that messages have TOO MUCH crust? That's UX issue, if anything. I think it's safe to postpone.

```
FORMAT = "%(levelname)s %(asctime)s %(name)s:%(lineno)d: %(message)s"
```

File: src/instructlab/configuration.py [hunk 16]
```diff
@@ -200,7 +201,10 @@ class _general(BaseModel):
     model_config = ConfigDict(extra="ignore")
 
     # additional fields with defaults
+    # TODO(leseb): move these to a separate class? so we can have a
```
[the remark this answers is not in the document]
Comment (leseb) [reply src/instructlab/configuration.py:204]: I just felt that we could have gone more granular with something like `cfg.general.log.{format,level}`.

File: scripts/functional-tests.sh [hunk 17]
```diff
@@ -132,6 +134,29 @@ wait_for_server(){
     fi
 }
 
+# It looks like GitHub action MacOS runner does not have graphics
```
Comment (danmcp) [inline scripts/functional-tests.sh:137]: What was the reason for this change?

Comment (leseb) [reply scripts/functional-tests.sh:137]: I made the check into its own function given that the `cleanup()` overwrites the config file now, this avoids the back-and-forth `sed` (to revert) within functions and inside the cleanup. Since we now apply this change again during cleanup.

Comment (danmcp) [reply scripts/functional-tests.sh:137]: Thanks!

**Review by danmcp — APPROVED** [review]

File: scripts/functional-tests.sh [hunk 18]
```diff
@@ -41,16 +41,18 @@ cleanup() {
...
-    local original_model_name='merlinite-7b-lab-Q4_K_M'
-    local original_model_filename="${original_model_name}.gguf"
+    local original_model_filename="merlinite-7b-lab-Q4_K_M.gguf"
     local temporary_model_filename='foo.gguf'
-    sed -i.bak "s/baz/${original_model_name}/g" "${ILAB_CONFIG_FILE}"
     mv "${ILAB_CACHE_DIR}/models/${temporary_model_filename}" "${ILAB_CACHE_DIR}/models/${original_model_filename}" || true
-    rm -f "${ILAB_CONFIG_FILE}.bak" serve.log "${ILAB_CACHE_DIR}/models/${temporary_model_filename}"
+    rm -f serve.log "${ILAB_CACHE_DIR}/models/${temporary_model_filename}"
+
+    # reset config file and re-init defaults
```
Comment (booxter) [inline scripts/functional-tests.sh:51]: you could have a separate function for init (`function init()`) and then you could reuse it below in lines 609-616?)

**Review by booxter — APPROVED** [review]

File: scripts/functional-tests.sh (hunk 18 above)
Comment (leseb) [reply scripts/functional-tests.sh:74]: Added thanks.

**Review by booxter — APPROVED** [review]
And thanks for removing the carriage pipe thing and replacing it with non-interactive.

**Comment by leseb** [comment]
I've decided to move forward with this since it's been over two months without consistent attention. I now have two approvals, so let's proceed with the merge. Thanks!
